railslink/railslink
# Admins Slash Command

This is a more consistent solution than my first attempt (#88) for making an admins slash command so folks can get a randomized list of admins if they need to report issues.

This relies on sending the slack event type via url params in the slash command configuration, which honestly seemed like a great way to handle using the same slack post event handling for both Events API requests and Slash Command API requests, but I could be argued out of it if necessary 😂 

This takes care of the feedback I got on the first go-round:

- Does not make an additional API call to get the admins, instead calls the scope on the SlackUser model
- Re-implements `publish_event` and `verify_token!` for admins slash command (first, because we want to verify all commands, derp, and second for consistency of how we're interacting with these event hooks - publishing the event to REDIS allows for us to look at the logs for troubleshooting better as recommended by @phallstrom 
- Removes unnecessary route access; the two requests (events API and slash commands API) have a different payload format but given that we're handling them the same, I'd rather accommodate the different params rather than set up a separate class for parsing different payload types for action

Additionally, I've added specs for this behavior, and updated documentation to include info about using the slash command (also since this is tangentially related to the COC I updated the welcome text to also include a link to that).

### Testing

If you want to test locally, you need to have `railslink-dev` in your local slack instance, and you will need to have `ngrok` installed
- I turned off whiny requests in `development.rb` because it doesn't like that you're sending external requests to a localhost and it will yell at you for _daring_ to expect console output: `config.web_console.whiny_requests = false` (see https://github.com/rails/web-console#configweb_consolewhiny_requests for more info)
- `ngrok http 3000`
- `rails s`
- navigate to `https://api.slack.com/apps/AEW9DAMFZ/slash-commands` and click `edit` next to `/admins`
- change the request url to: `https://[your ngrok url]/slack/event/?slack[event][type]=admins` and save your changes
- go to railslink-dev and type `/admins`
- you should get a shuffled list of admins in your local db; feel free to add some users to test the scope and shuffling
